co-log/src/Colog/Concurrent.hs

```diff
@@ -0,0 +1,129 @@
+-- |
+-- Sometimes an application may need to dump logs asynchonously.
+-- This may happen if the target where logs are dumped is
+-- slow and logs should be dumped in chunks, or if application
+-- should split CPU and I/O-bound tasks. In these cases it
+-- may make sense to run asynchronous logger.
+--
+-- Asynchronous logger runs in background thread, so users
+-- thread may dump logs asynchronously.
+--
+-- This module provides basic interface for the concurrent
+-- logging that may be either used as-is or can be used to
+-- build more complex concurrency abstractions.
+module Colog.Concurrent
+       ( -- $worker-thread
+         mkBackgroundThread
+       , runInBackgroundThread
+       -- $background-worker
+       , BackgroundWorker
+       , forkBackgroundLogger
+       , killBackgroundLogger
+       , convertToLogAction
+       ) where
+
+import Control.Concurrent
+import Control.Concurrent.STM.TBQueue
+
+import Colog.Core.Action (LogAction (..))
+import Colog.Concurrent.Internal
+
+-- $worker-thread
+-- Worker thread approach allows to create a single worker
+-- thread that will perform all logging. This approach allows
+-- user to use different loggers and change them on the flight.
+--
+-- With bracket resource control style code may look as
+-- following:
+--
+-- @
+-- bracket ('mkBackgroundThread' 4096)
+--         ('killBackgroundThread')
+--         $ \thread -> 
+--            let la1 = runInBackgroundThread worker logger1
+--                la2 = runInBackgroundThread worker logger1
+--            unLogger la1 msg1
+--            unLogger la2 msg2
+--            ..
+-- @
+--
+-- This code will run a background thread and process different
+-- loggers in that thread.
+--
+-- You may extend the logger and entire logger will be run
+-- on the background thread. It is not a case with
+-- 'forkBackgroundLogger'.
+
+-- | Create a background worker with a given capacity.
+-- If capacity is reached, then the thread that tries to
+-- write logs will be blocked.
+mkBackgroundThread :: Int -> IO (BackgroundWorker (IO ()))
+mkBackgroundThread capacity = do
+  queue <- newTBQueueIO (fromIntegral capacity)
```

Comment:
I see that `newTBQueueIO` function takes `Natural` in `stm-2.5.0.0`. Maybe it's a good idea to have `stm ^>= 2.5.0.0` in `build-depends` in that case and pass `Natural` to the `mkBackgroundThread` function as well? Also, probably it's not enough to have `stm >= 2.4` because `stm-2.4.5.1` contains very important bugfix for `TBQueue`:

* https://github.com/haskell/stm/issues/16

I even expect from GHC to give you warning about `fromIntegral` usage after you change type from `Int` to `Natural` because we have `-Widentities` warning enabled for `co-log`.

Reply:
It's up to you. As far as I understand `stm-2.5.0.0` will cut many older GHC's, in the case if you want to support 3 major versions then it's not an option. And I find it's easer to keep the same interface for all versions then doing CPP magic to provide the best option for each version.